File: src/perfects.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#define TOTAL 61

bool is_prime(long);
unsigned long long int is_perfect(int);

int main()
{
    unsigned long long int value;
    printf("p \t Number\n\n");
    for (int i = 1; i < TOTAL + 1; i++)
        if ((value = is_perfect(i)) != 0)
            printf("%d \t %llu\n", i, value);
}

bool is_prime(long n)
{
    if (n == 2) return true;

    if (n <= 1 || n % 2 == 0) return false;

    long maxdiv = sqrt(n) + 1;

    for (long i = 3; i < maxdiv; i += 2)
        if (n % i == 0)
            return false;

    return true;
}

unsigned long long int is_perfect(int a)
{
    if (is_prime(a)) {
        unsigned long long int aux = powl(2, a) - 1;
        if (is_prime(aux))
            return aux * powl(2, a - 1);
    }

    return 0;
}

File: src/perfects.py
from multiprocessing import Pool
TOTAL = 31


def is_prime(n: int) -> bool:

    if n == 2:
        return True

    if n <= 1 or n % 2 == 0:
        return False

    md: int = int(n ** 0.5) + 1

    for i in range(3, md, 2):
        if n % i == 0:
            return False

    return True


def is_perfect(a: int) -> int:

    if is_prime(a):
        aux = (2**a) -1

        if is_prime(aux):
            return aux*2**(a-1)
    return 0


if __name__=="__main__":

    print("p\tnumber\n")
    with Pool() as p:
        log = p.map(is_perfect, range(1, TOTAL + 1))

    for i, n in enumerate(log):
        if n:
            print(i, "\t", n)
        else:
            pass
